big_tests/tests/sasl_external_SUITE.erl

```diff
@@ -0,0 +1,136 @@
+    UserSpec = generate_user(C, "john"),
+    {ok, Client, _} = escalus_connection:start(UserSpec),
+
+    escalus_connection:stop(Client).
+
+no_cert_fails_to_authenticate(_C) ->
+    UserSpec = [{username, <<"no_cert_user">>},
+		{server, <<"localhost">>},
+		{password, <<"break_me">>},
+		{resource, <<>>}, %% Allow the server to generate the resource
+		{auth, {escalus_auth, auth_sasl_external}},
+		{starttls, required}],
+
+    Result = escalus_connection:start(UserSpec),
+    ?assertMatch({error, {connection_step_failed, _, _}}, Result),
+    {error, {connection_step_failed, _Call, Details}} = Result,
+    ?assertMatch({auth_failed, _, #xmlel{name = <<"failure">>}}, Details),
+    ok.
+
+
+generate_user(C, User) ->
+    generate_user(C, User, #{}).
+
+generate_user(C, User, TemplateValuesIn) ->
+    SSLDir = filename:join([path_helper:repo_dir(C), "tools", "ssl"]),
+
+    ConfigTemplate = filename:join(?config(data_dir, C), "openssl-user.cnf"),
+    {ok, Template} = file:read_file(ConfigTemplate),
+    TemplateValues = prepare_template_values(User, TemplateValuesIn),
+    OpenSSLConfig = bbmustache:render(Template, TemplateValues),
+    UserConfig = filename:join(?config(priv_dir, C), User ++ ".cfg"),
+    file:write_file(UserConfig, OpenSSLConfig),
+    UserKey = filename:join(?config(priv_dir, C), User ++ "_key.pem"),
+    UserCsr = filename:join(?config(priv_dir, C), User ++ ".csr"),
+
+    Cmd = ["openssl", "req", "-config", UserConfig, "-newkey", "rsa:2048", "-sha256", "-nodes",
+	   "-out", UserCsr, "-keyout", UserKey, "-outform", "PEM"],
+    {done, 0, _Output} = erlsh:run(Cmd),
+
+    UserCert = filename:join(?config(priv_dir, C), User ++ "_cert.pem"),
+    SignCmd = filename:join(?config(data_dir, C), "sign_cert.sh"),
+    Cmd2 = [SignCmd, "--req", UserCsr, "--out", UserCert],
+    LogFile = filename:join(?config(priv_dir, C), User ++ "singing.log"),
+    {done, 0, _} = erlsh:run(Cmd2, LogFile, SSLDir),
+
+    [{username, list_to_binary(User)},
+     {server, <<"localhost">>},
+     {password, <<"break_me">>},
+     {resource, <<>>}, %% Allow the server to generate the resource
+     {auth, {escalus_auth, auth_sasl_external}},
+     {ssl_opts, [{certfile, UserCert},
+		 {keyfile, UserKey}]},
+     {starttls, required}].
+
+
+prepare_template_values(User, TemplateValues) ->
+    Defaults = #{"cn" => User,
+		 "xmppAddrs" => ""},
+    XMPPAddrs = maybe_prepare_xmpp_addresses(maps:get("xmppAddrs", TemplateValues, undefined)),
+    maps:merge(Defaults, TemplateValues#{"xmppAddrs" => XMPPAddrs}).
+
+maybe_prepare_xmpp_addresses(undefined) ->
+    "";
+maybe_prepare_xmpp_addresses(Addrs) when is_list(Addrs) ->
+    AddrsWithSeq = lists:zip(Addrs, lists:seq(1, length(Addrs))),
+    Entries = [make_xmpp_addr_entry(Addr, I) || {Addr, I} <- AddrsWithSeq],
+    string:join(Entries, "\n").
+
+make_xmpp_addr_entry(Addr, I) ->
+    "otherName." ++ integer_to_list(I) ++ " = 1.3.6.1.5.5.7.8.5;UTF8:" ++ Addr.
```

Comment:
It may be useful for future maintainers to tell in a comment where this magic string comes from.